File: 0016-3sum-closest/0016-3sum-closest.java
class Solution {
    public int threeSumClosest(int[] nums, int target) {
        Arrays.sort(nums); 
        int closest = nums[0]+nums[1]+nums[2]; 

        for( int i = 0 ; i< nums.length-2 ; i++){
            int low = i+1; 
            int high = nums.length-1; 

            while( low< high){
                int sum = nums[i]+nums[low]+nums[high]; 

                if( Math.abs(sum-target) < Math.abs(closest-target) ){
                    closest = sum; 
                }
                if( sum < target) low++; 
                else high--; 
            }
        }
        return closest;  
    }
}

File: 0016-3sum-closest/0016-3sum-closest.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());  // Necessary for two-pointer approach

        int closestSum = nums[0] + nums[1] + nums[2];

        for (int i = 0; i < nums.size() - 2; i++) {
            int low = i + 1;
            int high = nums.size() - 1;

            while (low < high) {
                int currentSum = nums[i] + nums[low] + nums[high];

                if (currentSum == target) {
                    return currentSum;
                }

                // Update closest sum if needed
                if (abs(currentSum - target) < abs(closestSum - target)) {
                    closestSum = currentSum;
                }

                // Move pointers based on comparison
                if (currentSum < target) {
                    low++;
                } else {
                    high--;
                }
            }
        }
        return closestSum;
    }
};
